### PR TITLE
SPLICE-1882 

### PR DESCRIPTION
**Pull request by jleach4**
(1) Remove Automatic Clone After Scan for Row and RowLocation.
(2)  Fixes merge join to not automatically clone unless needed.  
(3) Remove object creation on RegionDataScanner for HCell.  
(4) Remove automatic cloning out of projection restrict operations.

### CODE REVIEW

**Comment by jleach4** [comment]
I ran IT's via Jenkins branch builder for cdh5.8.3.

File: hbase_sql/src/test/java/com/splicemachine/olap/OlapClientTest.java [hunk 1]
```diff
@@ -45,6 +45,7 @@
  * Created by dgomezferro on 3/17/16.
  */
 @SuppressWarnings("unused")
+@Ignore
```
Comment (snowspeeder) [inline hbase_sql/src/test/java/com/splicemachine/olap/OlapClientTest.java]: do we need to ignore or forgot to take this out?

Comment (jleach4) [reply hbase_sql/src/test/java/com/splicemachine/olap/OlapClientTest.java]: I ignored it since it fails sometimes without explanation.

Comment (snowspeeder) [reply hbase_sql/src/test/java/com/splicemachine/olap/OlapClientTest.java]: jenkins runs have not shown the issue recently maybe we should keep it running

File: splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/MergeAntiJoinIterator.java [hunk 2]
```diff
@@ -44,6 +44,10 @@ public MergeAntiJoinIterator(Iterator<ExecRow> leftRS,
     public boolean internalHasNext() {
         try {
             while (leftRS.hasNext()) {
+                if (left == null)
+                    left = leftRS.next().getClone();
+                else
+                    left.transfer(leftRS.next());
                 left = leftRS.next();
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/MergeAntiJoinIterator.java]: This line should be removed. Given that ITs pass I guess this is not covered by any test, could you write a quick one John?

**Comment by jleach4** [comment]
Per Daniel's comment:   JoinOperationIT references another ticket for fixing merge, bcast, etc. for antijoin...  @Ignore("DB-5173 no valid plan execution for sortmerge, bcast and merge")

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFunction;
+import scala.Tuple2;
+
+import java.util.*;
+
+/**
+ * Created by jleach on 8/30/17.
+ */
+public class CloneFirstFunction implements PairFunction<Tuple2<ExecRow,ExecRow>,ExecRow,ExecRow > {
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java]: Why is this needed? I don't understand very well what it's doing.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/AbstractMergeJoinIterator.java [hunk 4]
```diff
@@ -83,25 +88,42 @@ private int compare(ExecRow left, ExecRow right) throws StandardException {
...
+                int comparison = compare(left, rightRS.peek());
+                if (comparison == 0) { // if matches left, add to buffer
+                    currentRights.add(rightRS.next().getClone());
+                } else if (comparison < 0) { // if is greater than left, stop
+                    break;
+                } else {
+                    // if is less than left, read next right
+                    rightRS.next();
+                }
             }
```
Comment (yxia92) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/AbstractMergeJoinIterator.java]: If left is the same as the next leftRow, we should have the right dataset already prepared. Why do we need to compute again?

Comment (yxia92) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/AbstractMergeJoinIterator.java]: Please ignore this comment. My understanding was wrong. I think currentRights is prepared the second time the same left row key is seen.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ResultColumnList.java [hunk 5]
```diff
@@ -3800,5 +3586,278 @@ public void computeDistinctCardinality(CostEstimate costEstimate) throws Standar
...
+     * </p>
+     *
+     * This is the method that does the work.
+     */
+    void generateEvaluatedRow
+    (
+            ExpressionClassBuilder acb,
+            MethodBuilder userExprFun,
+            boolean genNulls,
+            boolean forMatchingClause
```
[the remark this answers is not in the document]
Comment (jleach4) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ResultColumnList.java]: Latest Derby Code line...

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java (hunk 3 above)
Comment (jleach4) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java]: The way the key by logic works it will never give you a chance to clone.  Basically the internal spark code flow is stick in map if not exists or in map, run reduce function with existing value.   

For example (key,value),

(1,1),
(1,2),
(1,3)
(2,1)
...

The reduce by key starts with ((1,1),(1,2)).  I am already screwed since I needed to clone (1,1) before it got into the map for reduceByKey.  Since I am not cloning all over my value becomes (1,1),(1,1) since they are referring to the same DVD.

Comment (dgomezferro) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/CloneFirstFunction.java]: I see, thanks for the explanation!
